Adding rolling location log appender

CuriousVini opened the pull request:
Adding custom HDFS rolling log appender using Logback apis. This includes `RollingLocationAppender` with `FixedWindowRollingPolicy` along with `SizeBasedTriggeringPolicy`.

Remaining:
- Compression after rollover 

Limitation:
- Prudent mode is not supported in this appender

Will create a separate PR for remaining work. 

Going to add unit tests for file permissions and if the log file already exists

Build:  http://builds.cask.co/browse/CDAP-RUT480-14

CuriousVini committed:
Adding rolling log appender for Location using logback apis

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,229 @@
+        rollover(logLocationIdentifier, eventObject);
+        OutputStream locationOutputStream = locationManager.getLocationOutputStream(logLocationIdentifier, fileName);
+        setOutputStream(locationOutputStream);
+        writeOut(eventObject);
```
`writeOut` is in FileAppender gets a lock on the file before writing. not sure if we need to do that.

CuriousVini replied:
That only happens in `prudent` mode. Besides, HDFS does not support locks. Also, since  the assumption is there is only one appender in single `TwillRunnable`, we do not need locks

gsps1 replied:
can't you write it out yourself, what's the advantage of using `writeOut`

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/SizeBasedTriggeringPolicy.java:
```diff
@@ -0,0 +1,107 @@
+    setMaxFileSize(maxFileSize);
+  }
+
+  private InvocationGate invocationGate = new InvocationGate();
```
is this needed?

CuriousVini replied:
This was part of logback SizeBasedTriggeringPolicy, I think it is used to avoid checking `activeLocation.length()` for every event.

gsps1 replied:
you are managing the size logic yourself, don't think you need this. what do you think?

CuriousVini replied:
right, fixed in another PR

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/SizeBasedTriggeringPolicy.java:
```diff
@@ -0,0 +1,107 @@
+  private InvocationGate invocationGate = new InvocationGate();
+
+  @Override
+  public boolean isTriggeringEvent(final File activeFile, final ILoggingEvent event) throws LogbackException {
```
since this is a custom class with `setLocation` can't you just have a `isTriggeringEvent(ILoggingEvent event)` method instead of having file and passing a dummy file?

CuriousVini replied:
good point. Done

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/SizeBasedTriggeringPolicy.java:
```diff
@@ -0,0 +1,107 @@
+    this.maxFileSize = FileSize.valueOf(maxFileSize);
+  }
+
+  long toFileSize(String value) {
```
unused ?

CuriousVini replied:
removed

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,229 @@
+        rollingPolicy.rollover();
+
+        // clear caches so new location gets created
+        locationManager.closeOutputStream(locationManager.getActiveFiles().get(logLocationIdentifier));
```
closing of output stream should happen before rollover. also its better to use the pattern of having LogOutputStream and that handle the case of closing output stream and evicting entry from map, etc. you can see `CDAPLogAppender` for reference.

CuriousVini replied:
fixed

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,229 @@
+  }
+
+  protected void encoderInit(OutputStream outputStream) {
+    if (encoder != null) {
```
encoders are specified via configuration ?

CuriousVini replied:
logback should provide it. In `start()` we do verify it

gsps1 reviewed (changes requested):
reviewed RollingLocationLogAppender, will look at others.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationIdentifier.java:
```diff
@@ -0,0 +1,62 @@
+      return false;
+    }
+
+    LocationIdentifier that = (LocationIdentifier) o;
```
can do `return (Objects.equals(o.namespaceId, that.namespaceId) && Objects.equals(appId, ..))`

CuriousVini replied:
done

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationIdentifier.java:
```diff
@@ -0,0 +1,62 @@
+  public int hashCode() {
+    int result = namespaceId.hashCode();
+    result = 31 * result + applicationId.hashCode();
+    return result;
```
similar, use `Objects.hash(...)`

CuriousVini replied:
done

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,191 @@
+      addInfo("No compression will be used");
+      compressionMode = CompressionMode.NONE;
+    }
+  }
```
better to remove compression related code, if its not used currently.

CuriousVini replied:
done

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,191 @@
+        Location deleteLocation = activeParentLocation.append(fileName);
+        if (deleteLocation.exists()) {
+          deleteLocation.delete();
+        }
```
I know this code is from logback, but do you have tests for these ?

CuriousVini replied:
Which test are you pointing at? We do have test for `rollover`

gsps1 replied:
tests for if things get deleted after max index.

CuriousVini replied:
That is already added as part of test right? At a time we should only have maxIndex number of files when we have many log events

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,191 @@
+
+  public FixedWindowRollingPolicy() {
+    minIndex = 1;
+    maxIndex = 7;
```
these should be configurable ?

CuriousVini replied:
It is configurable from logback

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,118 @@
+  public LocationManager(LocationFactory locationFactory) {
+    this.logDirLocation = locationFactory.create("plugins/logs");
+    this.activeFiles = new HashMap<>();
+    this.activeFilesToLocation = new HashMap<>();
```
you don't need two maps, just use one.

CuriousVini replied:
Since each `Location#getOutputStream` creates a new map, we are using 2 maps. Is there a way to get Location out of OutputStream?

gsps1 replied:
Location has `getOutputStream`

CuriousVini replied:
I dont think we have a way to get Location out of provided `OutputStream`

CuriousVini replied:
right. done

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,118 @@
+  }
+
+
+  protected LocationIdentifier getLocationIdentifier(Map<String, String> propertyMap) throws IllegalArgumentException,
```
can be package private right ? as there's no one extending this class. doesn't have to be protected.

CuriousVini replied:
done

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,118 @@
+    return new LocationIdentifier(namespaceId, application);
+  }
+
+  protected OutputStream getLocationOutputStream(LocationIdentifier locationIdentifier, String fileName)
```
add comments about the fileName structure, etc.

CuriousVini replied:
added it in test logback

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,118 @@
+    return logDirLocation.append(locationIdentifier.getNamespaceId()).append(locationIdentifier.getApplicationId());
+  }
+
+  protected void ensureDirectoryCheck(Location location) throws IOException {
```
private

CuriousVini replied:
removed this method instead using `Locations.mkdirsIfNotExists`

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,118 @@
+    }
+  }
+
+  public void close() {
```
public ?  if this class implements `Closeable` then this can be public.

CuriousVini replied:
Why?

CuriousVini replied:
now `LocationManager` implements `Flushable` and `Closeable`

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,229 @@
+
+  @Override
+  public void flush() throws IOException {
+    // no-op
```
no-op ? so when do you sync to file ?

CuriousVini replied:
`outputStream.flush()` is called on every `encoder#doAppend()` and `encoder#stop()` call. Is that what you mean by `sync`?

gsps1 replied:
for HDFS we want to call `hsync` by checking if the outputStream is of type `Syncable`. 
Also in PatternLayoutEncoder only when `immediateFlush` is enabled for the encoder, flush would be called on every append. else it wouldn't be. do you know when the flush will be called in that scenario ? And what if the encoder logic changes in logback in the future, we can't rely on them to flush right ?

CuriousVini replied:
hmm. For this, I have overridden `OutputStreamAppender#setLayout` and did `immediateFlush = false`. This means, wrapped encoder used to append log events will no longer be flushing on every log event. Instead, flush can be applied when `Appender#flush()` is called.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/SizeBasedTriggeringPolicy.java:
```diff
@@ -0,0 +1,107 @@
+  /**
+   * The default maximum file size.
+   */
+  public static final long DEFAULT_MAX_FILE_SIZE = 10 * 1024 * 1024; // 10 MB
```
weird that this is long, while maxFileSize is of type `FileSize`

CuriousVini replied:
why? It gets converted to `FileSize` in `setMaxFileSize`

gsps1 reviewed (changes requested):
few more comments

gsps1 commented:
@CuriousVini I don't see TimeBasedRollingPolicy implementation. we need that.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationInfo.java:
```diff
@@ -0,0 +1,50 @@
+ * Location information used by {@link LocationManager}
+ */
+public class LocationInfo {
+  private Location location;
```
final variables

CuriousVini commented:
@gsps1 about `TimeBasedRollingPolicy ` since this is a sample implementation for custom appender, I can add it in a separate PR if needed, let me know if you think otherwise

CuriousVini commented:
@gsps1 thanks for the review, I have addressed your comments. Please take another pass when you get a chance. Also I have added file permission related code. I will add unit tests for it as a separate commit.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,151 @@
+  public void rollover() throws RolloverFailure {
+    Location parentLocation = Locations.getParent(activeFileLocation);
+
+    if (parentLocation == null) {
```
when would this be null ?

CuriousVini replied:
`Locations.getParent()` has `Nullable` return type in case location is  root `/`.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,151 @@
+        String fileName = fileNamePattern.convertInt(maxIndex);
+        Location deleteLocation = parentLocation.append(fileName);
+        if (deleteLocation.exists()) {
+          deleteLocation.delete();
```
add a log here,`Log.debug("deleting oldest file {}", deleteLocation)`

CuriousVini replied:
done

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,151 @@
+          }
+        }
+      } catch (IOException e) {
+        throw new RolloverFailure(e.getMessage());
```
can we also include the actual exception, rather than just message. I understand that this is copied from logback, but we shouldn't skip exceptions.

CuriousVini replied:
`rollover` method from logback only throws `RolloverFailure`. It only has a constructor which takes string. I have added a method to convert stack trace of an exception to string.

chtyim replied:
You can do 

```java
RolloverFailure f = new RolloverFailure(e.getMessage());
f.addSuppressed(e);
throw f;
```

to kind of preserve the information (and stacktrace) of the original failure.

CuriousVini replied:
I see, fixed

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,151 @@
+        closeable.close();
+        activeFileLocation.renameTo(parentLocation.append(fileNamePattern.convertInt(minIndex)));
+      } catch (IOException e) {
+        throw new RolloverFailure(String.format("Exception while renaming file: %s, %s",
```
what if the exception was from `closeable.close`, also if rename fails, the old file will exist as active location but it would be closed already?, what would happen in that scenario ?

CuriousVini replied:
In current implementation:

1.) When `closeable.close()` fails:
We remove the file outputstream from `activeFiles` after closing the outputstream. So if there was an error in closing the outputstream, we catch that exception. So the rollover will not happen. But we can still write to that file because outputstream is not closed.

2.) When `activeFileLocation.renameTo` fails:
We remove the active file from `activeFiles` map after closing the outputstream. So if `activeFileLocation.renameTo` fails, we still write to the same file as if the file to be written exists.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/SizeBasedTriggeringPolicy.java:
```diff
@@ -0,0 +1,73 @@
+    invocationGate.updateMaskIfNecessary(now);
+
+    try {
+      return (activeLocation.length() >= maxFileSize.getSize());
```
i believe terence suggested to keep count of the size of the file in memory than using hdfs to get size of file, as that could be inaccurate

CuriousVini replied:
right, while testing I was not able to get correct size of file. So, I have fixed it to keep track of number of bytes written in memory for a given outputstream.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,274 @@
+    final Location logFileLocation = locationOutputStream.getLocation();
+    if (triggeringPolicy instanceof LocationTriggeringPolicy) {
+      ((LocationTriggeringPolicy) triggeringPolicy).setLocation(logFileLocation);
+      if (((LocationTriggeringPolicy) triggeringPolicy).isTriggeringEvent(eventObject)) {
```
also there were discussion around if we should do this for every event as it could be expensive to calculate size. add comments around if we decided to do it for every event and why, etc.

CuriousVini replied:
that is part of TriggeringPolicy

CuriousVini replied:
removed those changes because now the comparison will not be expensive

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,149 @@
+   * Appends information from location identifier to logBaseDir
+   */
+  Location getLogLocation(LocationIdentifier locationIdentifier) throws IOException {
+    return logBaseDir.append(locationIdentifier.getNamespaceId()).append(locationIdentifier.getApplicationId());
```
i thought we decided to add `instanceId` in location too, right ?

gsps1 replied:
hmm, instance id seems to be part filename..

CuriousVini replied:
thats right, the structure of the file name is explained in sample `rolling-appender-logback-test.xml`

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,149 @@
+    Collection<LocationOutputStream> locations = new ArrayList<>(activeLocations.values());
+    for (LocationOutputStream locationOutputStream : locations) {
+      locationOutputStream.flush();
+    }
```
will the framework be the only person calling flush ? since you are using logback's encoder they could call flush too (on the file) right? it that okay ?

CuriousVini replied:
That is part of framework. Also we have made ImmediateFlush to false

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,149 @@
+    Location logFile = getLogLocation(locationIdentifier).append(fileName);
+    Location logDir = Locations.getParent(logFile);
+    // check if parent directories exist
+    Locations.mkdirsIfNotExists(logDir);
```
what if a fie exists at the logDir and its not a directory ? this is handled in `CDAPLogAppender` directory check.

CuriousVini replied:
Its the same condition

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,149 @@
+        throw new IOException(String.format("Parent Directory for %s is null", logFile.toURI().toString()));
+      }
+      Location tempLocation = logFile.renameTo(logDir.append(Long.toString(now)));
+      logFile.createNew(filePermissions);
```
you should not create the file before checking if `renameTo` succeeded

gsps1 replied:
what happens if `createNew` fails?

CuriousVini replied:
Its handled by caller by throwing `LogbackException`

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,274 @@
+        LocationIdentifier logLocationIdentifier = locationManager.getLocationIdentifier(eventObject
+                                                                                           .getMDCPropertyMap());
+        rollover(logLocationIdentifier, eventObject);
+        OutputStream locationOutputStream = locationManager.getLocationOutputStream(logLocationIdentifier, filePath);
```
`filePath` becomes `fileName` in the LocationOutputStream ? that doesn't sound correct.

CuriousVini replied:
renamed `fileName` to `filePath`. The structure is explained in sample logback file: `<basePath>/namespaceId/applicationId/<filePath>`

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,149 @@
+        // this should never happen
+        throw new IOException(String.format("Parent Directory for %s is null", logFile.toURI().toString()));
+      }
+      Location tempLocation = logFile.renameTo(logDir.append(Long.toString(now)));
```
the new file name would just be timestamp ?

gsps1 replied:
oh, this is just temp file, that you would copy to the active location output stream please add comments, and todo's for potential size limits.

CuriousVini replied:
done

CuriousVini replied:
Added comments

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,149 @@
+      try (BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(tempLocation.getInputStream()))) {
+        String line;
+        while ((line = bufferedReader.readLine()) != null) {
+          outputStream.write(line.getBytes());
```
you should use `Locations.newInputSupplier` for getting input stream for a location.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,149 @@
+        }
+      }
+      // delete temporary file
+      tempLocation.delete();
```
what is this step fails? the new location will be open, but active locations wouldn't have been updated. It might be better to retry this delete than throw exception.

CuriousVini replied:
Hmm, good point. fixed it by retrying 5 times and sleep time is 100 Millisecs.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,149 @@
+    return new LocationIdentifier(namespaceId, application);
+  }
+
+  OutputStream getLocationOutputStream(final LocationIdentifier locationIdentifier, String fileName)
```
add docs for this method. not clear whats the expectation/format for `fileName`is.

CuriousVini replied:
done

chtyim replied:
I don't think the `locationIdentifier` needs to be `final`.

CuriousVini replied:
fixed

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,274 @@
+              }
+            });
+
+            ((LocationRollingPolicy) rollingPolicy).rollover();
```
this casting isn't necessary.

CuriousVini replied:
done

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,274 @@
+
+  @Override
+  public void stop() {
+    if (encoder != null) {
```
when will the encoder be null ?

CuriousVini replied:
when its not passed from logback

gsps1 reviewed (changes requested):
few comments, mainly looked at RollingLocationLogAppender and LocationManager, can you think about exception scenarios and how that will be handled, adding comments around how that's handled, etc.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,274 @@
+        rollover(logLocationIdentifier, eventObject);
+        OutputStream locationOutputStream = locationManager.getLocationOutputStream(logLocationIdentifier, filePath);
+        setOutputStream(locationOutputStream);
+        writeOut(eventObject);
```
so `setOutputStream` returns if encoder is not set (null), however writeOut just tries to write using the encoder and that would cause NPE right ?

also we have the output stream opened, even if encoder is not set, though encoder is the only person to write to output stream here.

CuriousVini replied:
When we start log appender we check if `encoder` is null or not

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,151 @@
+ * When rolling over, renames files according to a fixed window algorithm.
+ */
+public class FixedWindowRollingPolicy extends LocationRollingPolicyBase {
+  // TODO Support compression
```
Do we have a JIRA?

CuriousVini replied:
created: CDAP-8369. will add it to TODO

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,151 @@
+public class FixedWindowRollingPolicy extends LocationRollingPolicyBase {
+  // TODO Support compression
+  private int minIndex;
+  private static final String FNP_NOT_SET =
```
Have all static fields before non-static one.

CuriousVini replied:
done

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationOutputStream.java:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  public void setLocation(Location location) {
+    this.location = location;
```
Why have this class mutable? Immutable  class is better in general.

CuriousVini replied:
fixed

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationOutputStream.java:
```diff
@@ -0,0 +1,73 @@
+  @Override
+  public void close() throws IOException {
+    LOG.trace("Closing file {}", location);
+    this.flush();
```
No need

CuriousVini replied:
done

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationOutputStream.java:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Location outputstream used by {@link LocationManager}
+ */
+public class LocationOutputStream implements Flushable, Closeable {
```
If it is only used by `LocationManager` in the same package, this class should be package private.

chtyim replied:
Also, this class is not an `OutputStream`, so don't call it `OutputStream`.

chtyim replied:
On the other hand, you can actually make this class an `OutputStream` by extending from `FilterOutputStream`, which I think is a better approach

CuriousVini replied:
fixed

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,151 @@
+      maxIndex = minIndex;
+    }
+
+    if ((maxIndex - minIndex) > MAX_WINDOW_SIZE) {
```
What is window size? Is it number of files?

CuriousVini replied:
Yes that is max number of files allowed (window size). It is hard coded to be `12` even in logback implementation. I changed it to `20` as per https://logback.qos.ch/xref/ch/qos/logback/core/rolling/FixedWindowRollingPolicy.html. They do not encourage using large window sizes I guess its because for `FixedWindowRollingPolicy` we go over all the files and change file names.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,151 @@
+        // Delete the oldest file, to keep Windows happy.
+        String fileName = fileNamePattern.convertInt(maxIndex);
+        Location deleteLocation = parentLocation.append(fileName);
+        if (deleteLocation.exists()) {
```
Just call `delete()`, no need to call `exists` as it doesn't guarantee the file exists when `delete()` is actually called.

CuriousVini replied:
This needed because `fileName` is a path. I renamed it to `filePath` to avoid confusion

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,151 @@
+          String toRenameStr = fileNamePattern.convertInt(i);
+          Location toRename = parentLocation.append(toRenameStr);
+          // no point in trying to rename an non existent file
+          if (toRename.exists()) {
```
This check is also not necessary, the `renameTo` will return `null` if rename didn't happened (due to whatever reason).

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,149 @@
+  protected static final String TAG_NAMESPACE_ID = ".namespaceId";
+  protected static final String TAG_APPLICATION_ID = ".applicationId";
+
+  private Location logBaseDir;
```
final fields

CuriousVini replied:
done

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,149 @@
+    this.filePermissions = filePermissions;
+  }
+
+  LocationIdentifier getLocationIdentifier(Map<String, String> propertyMap) throws IllegalArgumentException,
```
`IllegalArgumentException` is a runtime exception, no need to declare it in `throws`. Also, it's good to have javadoc for non-private method to include what this method does, what's the param and what exceptions it will throw under what condition.

CuriousVini replied:
done

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,149 @@
+    IOException {
+
+    String namespaceId = propertyMap.get(TAG_NAMESPACE_ID);
+    Preconditions.checkArgument(propertyMap.containsKey(TAG_APPLICATION_ID),
```
Better just call `get` and check if `application == null`

CuriousVini replied:
done

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,149 @@
+  }
+
+  OutputStream getLocationOutputStream(final LocationIdentifier locationIdentifier, String fileName)
+    throws IOException {
```
Usually I prefer exception at the end of the method to make it not mixing up with the method body. You can do so by adding a new line before `String fileName`).

CuriousVini replied:
done

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    Location logFile = getLogLocation(locationIdentifier).append(fileName);
+    Location logDir = Locations.getParent(logFile);
```
No need to call `getParent()`... 

```java
Location logDir = getLogLocation(locationIdentifier);
Location logFile = logDir.append(fileName);
```

CuriousVini replied:
We get the parent because `filePath` is a path with directory of file: `security/logFile.log. Because `Locations.createNew()` creates a file, we are doing `getParent`

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,149 @@
+      // copy its contents to new file and delete old file
+      long now = System.currentTimeMillis();
+      // rename existing file to temp file
+      if (logDir == null) {
```
No need this check (since you shouldn't be using `Locations.getParent()` in this case.

CuriousVini replied:
see above comments

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,149 @@
+        throw new IOException(String.format("Can not rename file %s", logFile.toURI().toString()));
+      }
+
+      OutputStream outputStream = logFile.getOutputStream();
```
Add some comment to tell what is happening here.

CuriousVini replied:
done

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,149 @@
+      }
+
+      OutputStream outputStream = logFile.getOutputStream();
+      try (BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(tempLocation.getInputStream()))) {
```
First, when using `InputStreamReader`, also include a charset, otherwise the content may get messed up.
Second, in here, you shouldn't use `Reader` at all. File copying should be at byte level.

You can use `ByteStreams.copy` to help.

CuriousVini replied:
thanks, fixed

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,149 @@
+        throw new IOException(String.format("Can not rename file %s", logFile.toURI().toString()));
+      }
+
+      OutputStream outputStream = logFile.getOutputStream();
```
If something goes wrong below, this `OutputStream` becomes a leaking resource

CuriousVini replied:
Hmm, right. Fixed it by wrapping it with try/catch

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationRollingPolicyBase.java:
```diff
@@ -0,0 +1,80 @@
+ */
+public abstract class LocationRollingPolicyBase extends ContextAwareBase implements LocationRollingPolicy {
+
+  protected CompressionMode compressionMode = CompressionMode.NONE;
```
Generally, better to have fields as `private` and expose getter for sub-classes to gain access to the field rather than using the field directly, especially these fields are non-final.

CuriousVini replied:
removed code related to compression

gsps1 replied:
this comment is still valid right ? you haven't made them private yet.

chtyim commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,274 @@
+    final Location logFileLocation = locationOutputStream.getLocation();
+    if (triggeringPolicy instanceof LocationTriggeringPolicy) {
+      ((LocationTriggeringPolicy) triggeringPolicy).setLocation(logFileLocation);
+      if (((LocationTriggeringPolicy) triggeringPolicy).isTriggeringEvent(eventObject)) {
```
What if you cast it once and store it in a new local variable instead of keep casting?

CuriousVini replied:
good point, done

CuriousVini commented:
@chtyim @gsps1 Thanks for the review! I have addressed comments and tested my changes on cluster for following test scenarios:

- Appends to correct location - basePath/ns/app/filePath
- Rollover works
- File rename works if the file already exists
- Files have correct file permissions - -rw-r--r--
	- Permissions on directory is not same as permissions on file

Could you please take another pass? I will integrate it with CDAP Log framework in a separate PR.

CuriousVini commented:
Build is green here: http://builds.cask.co/browse/CDAP-RUT480-7 @gsps1 @chtyim please take another pass when you get a chance, thanks

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,177 @@
+ */
+public class LocationManager implements Flushable, Closeable {
+  protected static final String TAG_NAMESPACE_ID = ".namespaceId";
+  protected static final String TAG_APPLICATION_ID = ".applicationId";
```
can use the constants in `Constants.Logging`

CuriousVini replied:
Will be addressed in another PR

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,177 @@
+      }
+
+      // delete temporary file, if it fails, retry 5 times, the file will be there on the disk after 5 retries
+      // TODO remove all the existing temp files
```
can you address this todo?

CuriousVini replied:
done

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,268 @@
+
+  // TODO CDAP-8196 LocationFactory should be passed from Log framework
+  @Inject
+  private LocationFactory locationFactory;
```
its being passed from appender context currently. can you update your code to use that?

gsps1 replied:
is this what you plan to address in next PR ?

CuriousVini replied:
Will address it in another PR

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,268 @@
+      }
+
+      String namespaceId = eventObject.getMDCPropertyMap().get(LocationManager.TAG_NAMESPACE_ID);
+      if (!namespaceId.equals(NamespaceId.SYSTEM.getEntityName())) {
```
better call `getNamespace()` as `getEntityName` calls that.

CuriousVini replied:
fixed in another PR

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,268 @@
+        return;
+      }
+
+      String namespaceId = eventObject.getMDCPropertyMap().get(LocationManager.TAG_NAMESPACE_ID);
```
will be good to check if `namespaceId` exists

gsps1 replied:
meant, non-null check.

CuriousVini replied:
will be fixed in another PR

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,164 @@
+              throw new RolloverFailure(String.format("Not able to rename file: %s", toRename.toURI().toString()));
+            }
+          } else {
+            LOG.trace("Skipping roll-over for inexistent file {}", toRenameStr);
```
it will be better to log `toRename` than `toRenameStr` as its a trace. `toRename` would be more useful when debugging

CuriousVini replied:
fixed in another PR

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,164 @@
+            Location newName = parentLocation.append(fileNamePattern.convertInt(i + 1));
+            // throw exception if rename fails, so that in next interation of rollover, it will be retried
+            if (toRename.renameTo(newName) == null) {
+              LOG.warn("Failed to rename location: {}", toRename.toURI().toString());
```
also log the from and to parts of rename.. `Failed to rename {} to {}`

gsps1 replied:
rename will overwrite by default right, if the destination already exists ?

CuriousVini replied:
1.) Fixed in another PR, 2.)rename overwrites if the file already exists

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,268 @@
+              @Override
+              public void close() throws IOException {
+                locationOutputStream.close();
+                locationManager.getActiveLocations().remove(logLocationIdentifier);
```
you can do remove from the activeLocation in a finally block. so that gets removed even if there was exception while closing the file.

CuriousVini replied:
If there is an exception in output stream close(), we dont want to remove outputstream, instead, we keep appending to open file.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,268 @@
+    if (encoder != null) {
+      try {
+        LOG.info("Stopping appender {}", this.name);
+        encoder.close();
```
would it be safer to do `encoder.init(null)` before calling close ? otherwise, encoder will close one of the output stream (latest one it inited) and we want to manage flushing/closing output streams.

CuriousVini replied:
fixed in another PR

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,268 @@
+      try {
+        LOG.info("Stopping appender {}", this.name);
+        encoder.close();
+        locationManager.close();
```
better to do this outside (before) this if condition

CuriousVini replied:
fixed in next PR

gsps1 commented on cdap-watchdog/src/test/resources/rolling-appender-logback-test.xml:
```diff
@@ -0,0 +1,63 @@
+    <basePath>/tmp/plugins/applogs</basePath>
+    <!--filePath uses instanceId which is a property set by LogFramework. We use it to avoid multiple log saver
+    instances writing to same file in case of kafka partition key reconfiguration supported by CDAP-8104-->
+    <filePath>securityLogs/logFile-${cdap.log.saver.instance.id}.log</filePath>
```
i believe this property is going to be gone in the next PR

CuriousVini replied:
Since log framework is not exposing `cdap.log.saver.instance.id`, I am still using this property, but replacing it with program Id in another PR

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationTriggeringPolicyBase.java:
```diff
@@ -0,0 +1,58 @@
+ */
+public abstract class LocationTriggeringPolicyBase extends ContextAwareBase implements LocationTriggeringPolicy {
+  protected Location activeLocation;
+  protected long activeLocationSize;
```
terence suggested to avoid protected usage and he suggested alternate approach right? can you make that change.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationOutputStream.java:
```diff
@@ -0,0 +1,79 @@
+ * Location outputstream used by {@link LocationManager} which holds location, number of bytes written to a location
+ * and its open outputstream
+ */
+public class LocationOutputStream extends FilterOutputStream {
```
why FilterOutputStream why not just `OutputStream`

CuriousVini replied:
Because it provides basic implementation to methods of `OutputStream`. See above comment from Terence

gsps1 reviewed (changes requested):
few comments.

CuriousVini commented:
@gsps1 Addressed comments. Please take another pass.

CuriousVini commented:
Build is green: http://builds.cask.co/browse/CDAP-RUT480-8

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,201 @@
+  private int maxIndex;
+
+  /**
+   * It's almost always a bad idea to have a large window size, say over 12.
```
update this comment. as we are using max window size of 20.

CuriousVini replied:
fixed

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,201 @@
+    if (maxIndex < minIndex) {
+      addWarn("MaxIndex (" + maxIndex + ") cannot be smaller than MinIndex ("
+                + minIndex + ").");
+      addWarn("Setting maxIndex to equal minIndex.");
```
do these status, get printed as logs ? where does it show up ?

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,201 @@
+    }
+
+    // Inside this method it is guaranteed that the hereto active log file is
+    // closed.
```
this comment is also not relevant right ? meaning, we are closing it on L#140

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationRollingPolicyBase.java:
```diff
@@ -0,0 +1,80 @@
+  protected Closeable closeable;
+  private boolean started;
+
+  public void setFileNamePattern(String fnp) {
```
add comments, saying `logback calls these methods, so do not remove`

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,229 @@
+  private final Location logBaseDir;
+  private Map<LocationIdentifier, LocationOutputStream> activeLocations;
+  private String filePermissions;
+  private String dirPermissions;
```
the above three can be final too

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,229 @@
+        // TODO: Handle existing file in a better way rather than copying it over
+        OutputStream outputStream = new LocationOutputStream(logFile, logFile.getOutputStream());
+        activeLocations.put(locationIdentifier, (LocationOutputStream) outputStream);
+        ByteStreams.copy(inputStream, outputStream);
```
`ByteStreams.copy` does not close or flush either of the stream. you have to close the output stream here.

gsps1 replied:
oh, here you don't want to close the output stream, as the appends will be using this output stream right?. you can just call flush on output stream in that case, after copy.

CuriousVini replied:
If we close output stream I can not use it further for append

gsps1 replied:
yeah, still need to call flush

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,229 @@
+        // clean up all the temp files for which deletion failed
+        tempLocation.delete();
+        for (Location location : logDir.list()) {
+          if (location.toURI().toString().contains("temp-")) {
```
better to do something like `location.getName().startsWith("temp")` now even if the dir path contains temp, all files will be matched.

gsps1 replied:
also can you move this logic to a method `deleteTempFiles(Location dir)`

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,229 @@
+        }
+      } catch (IOException e) {
+        // do not throw any exception while deleting temp directory
+        LOG.warn("Not able to delete temp location, will be retried later");
```
also add `e.getMessage()` to this

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -0,0 +1,229 @@
+   */
+  @Override
+  public void flush() throws IOException {
+    Collection<LocationOutputStream> locations = new ArrayList<>(activeLocations.values());
```
you don't have to make a copy of values for flush. for close we make a copy to avoid concurrentModificationException, as when we close the stream, we are remove the elements from the map of active locations.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationRollingPolicy.java:
```diff
@@ -0,0 +1,36 @@
+ */
+public interface LocationRollingPolicy extends RollingPolicy {
+  /**
+   /**
```
duplicate comment lines

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationTriggeringPolicy.java:
```diff
@@ -0,0 +1,43 @@
+   */
+  void setLocation(Location location);
+
+  void setActiveLocationSize(long size);
```
add javadocs for this too.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,253 @@
+          }
+        } catch (RolloverFailure e) {
+          // we do not want to stop processing because roll over failed. so catch it and process the event
+          LOG.warn("Attempt to rollover failed for appender {}.", name);
```
add `e.getMessage()` to the log.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,253 @@
+
+  private void rollover(final LocationIdentifier logLocationIdentifier, ILoggingEvent logEvent) {
+    if (!locationManager.getActiveLocations().containsKey(logLocationIdentifier)) {
+      LOG.info("Decided not to roll over for first event");
```
this info message is not useful

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -0,0 +1,253 @@
+            locationRollingPolicy.setLocation(locationOutputStream.getLocation(), new Closeable() {
+              @Override
+              public void close() throws IOException {
+                locationOutputStream.close();
```
do you want to close this quietly too, as you close output streams quietly when we stop ?

CuriousVini replied:
I think we should not close it quietly because if there is any exception in closing the stream, we should not rename the file.

gsps1 replied:
why ? if there was exception during close due to file system, then rename would fail with similar error, but now if there is an exception during close, the map would still have reference to this invalid output stream and subsequent writes will use the same output stream and keep failing ?

CuriousVini replied:
we catch the exception, so rename will not be called

chtyim replied:
If close fail, it means the outputstream could be in some bad state that may not be recoverable from the program space. So it's better not to keep writing to it. However, it does create a problem that the next event that come will try to open a new file with the same file name (assuming we don't rename or the rename failed as well), which is not good. I think the best sequence is this:

I think the best sequence is something like this:

1. Remove the stream from the map
2. Attempt to close the file. If it fails, remember the output stream and throw exception. The framework will log a warn and proceed.
3. If step 2 succeeded, try to rename the file. If rename failed, remember the location and throw exception. The framework will log a warn and proceed.
4. Next time when `append()` is called, if either step 2 or 3 was a failure, retry the close/rename based on the remembered stream/location before appending (hence creating new file). If this fail, throw exception. The framework will handle it by sending the same event again.
5. If step 4 cannot proceed, the framework will keep repeating step 4. It might result in filling up the in memory sort buffer, hence stop polling from Kafka. I think that's ok as that's the back pressure mechanism in the framework and it still guarantees there is no loss of data.

Note that the framework only has this mechanism in distributed mode, as we have Kafka to act as the backing buffer. In local mode, everything happen in memory. If there is a failure to Append, the event will simply be ignore (the chance that it fails in local mode is small anyway).

gsps1 reviewed (commented):
few more comments.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,211 @@
+          if (toRename.exists()) {
+            Location newName = parentLocation.append(fileNamePattern.convertInt(i + 1));
+            // throw exception if rename fails, so that in next iteration of rollover, it will be retried
+            if (toRename.renameTo(newName) == null) {
```
this could throw `IOException` too, we need to update processedIndex also in that case.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java:
```diff
@@ -0,0 +1,211 @@
+        // close outputstream of active location before renaming it
+        closeable.close();
+
+        if (activeFileLocation.renameTo(parentLocation.append(fileNamePattern.convertInt(minIndex))) == null) {
```
what happens when this rename fails ? don't think that's handled for retry. as closeable's close succeeding would have removed this from the active locations map, right ?

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -146,16 +153,40 @@ private void rollover(final LocationIdentifier logLocationIdentifier, ILoggingEv
+                } catch (IOException e) {
+                  // If there is an exception while closing the outputstream, remember it and throw an exception so
+                  // that it can be closed on another event append
+                  locationManager.getInvalidOutputStreams().put(identifier, locationOutputStream);
```
does this have to be a map ? there will be only one entry right ? as the last failed event will be kept retrying. can you just remember the last failed output stream in a variable ?

gsps1 reviewed (commented):
few comments

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -173,21 +171,17 @@ public void close() throws IOException {
+        // because close was successful make this output stream null
+        invalidOutputStream = null;
+      } catch (IOException e) {
+        LOG.warn("Exception while closing invalid output stream for {}, will retry to close it later",
```
log this as trace, as it gets retried and we will end-up logging a lot.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -173,21 +171,17 @@ public void close() throws IOException {
+      LocationOutputStream invalidOutputStream = locationManager.getInvalidOutputStream();
+      try {
+        invalidOutputStream.close();
+        // because close was successful make this output stream null
```
log a message in info here, saying close of the stream was successful.  will be useful.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -195,8 +194,12 @@ Location getLogLocation(LocationIdentifier locationIdentifier) throws IOExceptio
   @VisibleForTesting
-  Map<LocationIdentifier, LocationOutputStream> getInvalidOutputStreams() {
-    return invalidOutputStreams;
+  LocationOutputStream getInvalidOutputStream() {
```
add `@Nullable` annotation

gsps1 replied:
also this class is not used in tests, can remove the `visiblefortesting` annotation.

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java:
```diff
@@ -195,8 +194,12 @@ Location getLogLocation(LocationIdentifier locationIdentifier) throws IOExceptio
+    return invalidOutputStream;
+  }
+
+  public void setInvalidOutputStream(LocationOutputStream invalidOutputStream) {
```
doesn't have to be public. add `@nullable` annotation for param

gsps1 commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java:
```diff
@@ -173,21 +171,17 @@ public void close() throws IOException {
+      try {
+        invalidOutputStream.close();
+        // because close was successful make this output stream null
+        invalidOutputStream = null;
```
this is a local variable, you have to do `locationManager.setInvalidOutputStream(null)`

gsps1 reviewed (changes requested):
final comments.

gsps1 commented:
LGTM, please squash and merge.

CuriousVini commented:
Build is green. Merging
